Mathlib/RingTheory/Regular/Category.lean

```diff
@@ -0,0 +1,52 @@
+/-
+Copyright (c) 2025 Nailin Guan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jingting Wang, Wanyi He, Nailin Guan
+-/
+import Mathlib.Algebra.Homology.ShortComplex.ModuleCat
+import Mathlib.RingTheory.QuotSMulTop
+
+/-!
+# Categorical constructions for `IsSMulRegular`
+-/
+
+universe u v w
+
+variable {R : Type u} [CommRing R] (M : ModuleCat.{v} R)
+
+open CategoryTheory Ideal Pointwise
+
+/-- The short complex `M → M → M⧸xM` given by an element `x : R`. -/
+@[simps]
+def smulShortComplex (r : R) :
+    ShortComplex (ModuleCat R) where
+  X₁ := M
+  X₂ := M
+  X₃ := ModuleCat.of R (QuotSMulTop r M)
+  f := ModuleCat.ofHom (r • LinearMap.id)
+  g := ModuleCat.ofHom (r • (⊤ : Submodule R M)).mkQ
+  zero := by
+    ext m
+    simp only [ModuleCat.hom_comp, ModuleCat.hom_ofHom, LinearMap.coe_comp, Function.comp_apply,
+      LinearMap.smul_apply, LinearMap.id_coe, id_eq, Submodule.mkQ_apply,
+      ModuleCat.hom_zero, LinearMap.zero_apply, Submodule.Quotient.mk_eq_zero]
+    exact Submodule.smul_mem_pointwise_smul m r ⊤ trivial
+
+lemma smulShortComplex_function_exact (r : R) :
```

Comment:
This particular lemma should be stated using variables `(M : Type v) [AddCommGroup M] [Module R M]` instead.